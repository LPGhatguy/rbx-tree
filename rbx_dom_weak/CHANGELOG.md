# rbx\_dom\_weak Changelog

## Unreleased
* `RbxValueType` now derives `Serialize` and `Deserialize`
* Added support for `BrickColor`

## 1.4.0 (2019-03-27)
* Added support for new types:
	* `NumberRange`
	* [#15](https://github.com/LPGhatguy/rbx-dom/pull/15):
		* `Int64`
		* `Float64`
	* [#20](https://github.com/LPGhatguy/rbx-dom/pull/20):
		* `Rect`
		* `Ray`
		* `ColorSequence`
		* `NumberSequence`

## 1.3.0 (2019-03-14)
* Added `sort_children_by_key` and `sort_children_unstable_by_key` methods to `RbxInstance` to reorder children safely

## 1.2.0 (2019-03-13)
* `RbxTree` and `RbxInstance` are now clonable without an `unimplemented` panic
	* Cloned trees preserve their IDs as-is and there's no public API yet to transplant instances between trees. This is mostly useful for comparing trees before/after a mutation, which we're using in Rojo.

## 1.1.0 (2019-03-11)
* Marked `PhysicalProperties` as `pub`
* Fixed `Serialize` impl for `UnresolvedRbxValue`
	* This removes `Serialize` from `AmbiguousRbxValue`, which should not break any real code.

## 1.0.0 (2019-03-01)
* `RbxValue` and `RbxValueType` can no longer be matched exhaustively, which enables adding new types without breaking code in the future
* Added support for `UDim`, `UDim2`, and `Content` value types
* `PhysicalProperties` is no longer a stub type

## 0.3.1 (2019-02-26)
* Added support for bare bool values when deserializing `UnresolvedRbxValue`
* Implemented `Serialize` for `UnresolvedRbxValue` and `AmbiguousRbxValue`
* Implemented `From<RbxValue>` for `UnresolvedRbxValue`

## 0.3.0 (2019-02-14)
* Renamed crate from `rbx_tree` to `rbx_dom_weak`
* Added support for `Ref` values ([#8](https://github.com/LPGhatguy/rbx-dom/pull/8))
* Added `UnresolvedRbxValue` and `AmbiguousRbxValue`, intended to be used alongside `rbx_reflection` to make specifying values less verbose.

## 0.2.0 (2019-01-25)
* Added new variants for `RbxValue`:
	* Int32
	* Float32
	* Enum
	* Vector2
	* Color3
	* Color3uint8
	* Vector3int16
	* Vector2int16
	* CFrame
	* PhysicalProperties (Stub)

## 0.1.0
* Initial release
* Supports `String`, `Bool`, and `Vector3` property values