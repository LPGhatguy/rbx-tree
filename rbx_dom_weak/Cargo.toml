[package]
name = "rbx_dom_weak"
description = "Weakly-typed Roblox DOM implementation for Rust"
license = "MIT"
documentation = "https://docs.rs/rbx_dom_weak"
homepage = "https://github.com/LPGhatguy/rbx-dom"
repository = "https://github.com/LPGhatguy/rbx-dom.git"
version = "1.8.2"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[dependencies]
base64 = "0.10.1"
lazy_static = "1.3.0"
md5 = "0.6.1"
serde = "1.0"
serde_derive = "1.0"
serde_repr = "0.1"
uuid = { version = "0.7", features = ["v4", "serde"] }

[dev-dependencies]
serde_json = "1.0"
maplit = "1.0"
