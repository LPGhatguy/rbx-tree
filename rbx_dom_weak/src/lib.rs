//! rbx_dom_weak is a common representation of the Roblox DOM for Rust. It's
//! designed to play nicely with the borrow checker and allows accessing
//! instances by ID in constant time.
//!
//! Constructing a new tree of instances is accomplished by first creating an
//! [`RbxInstanceProperties`] object that describes the root instance of the
//! tree, and then wrapping it with an [`RbxTree`]:
//!
//! ```
//! use std::collections::HashMap;
//! use rbx_dom_weak::{RbxInstanceProperties, RbxTree};
//!
//! let props = RbxInstanceProperties {
//!     name: "My Cool Game".to_owned(),
//!     class_name: "DataModel".to_owned(),
//!     properties: HashMap::new(),
//! };
//!
//! let mut tree = RbxTree::new(props);
//! println!("ID of instance we just inserted is {}", tree.get_root_id());
//! ```
//!
//! Note that the [maplit] crate is incredibly useful for defining properties
//! inline.
//!
//! Once we have a tree, we can use [`RbxTree::insert_instance`] and
//! [`RbxTree::get_instance`] to add instances to the tree and retrieve them.
//!
//! ```
//! # use std::collections::HashMap;
//! # use rbx_dom_weak::{RbxInstanceProperties, RbxTree};
//! use rbx_dom_weak::RbxValue;
//! use maplit::hashmap;
//! #
//! # let props = RbxInstanceProperties {
//! #     name: "My Cool Game".to_owned(),
//! #     class_name: "DataModel".to_owned(),
//! #     properties: HashMap::new(),
//! # };
//! #
//! # let mut tree = RbxTree::new(props);
//! #
//! let http_service = RbxInstanceProperties {
//!     name: "HttpService".to_owned(),
//!     class_name: "HttpService".to_owned(),
//!     properties: hashmap! {
//          // Properties are represented via the RbxValue enum
//!         "HttpEnabled".to_owned() => RbxValue::Bool {
//!             value: true,
//!         },
//!     },
//! };
//!
//! let datamodel_id = tree.get_root_id();
//! let http_service_id = tree.insert_instance(http_service, datamodel_id);
//!
//! println!("HttpService has ID {}", http_service_id);
//! ```
//!
//! To change properties on an instance that's already present in the tree, use
//! [`RbxTree::get_instance_mut`]. Note that it isn't possible to add or remove
//! children through this method, use [`RbxTree::insert_instance`] instead.
//!
//! [`RbxTree`]: struct.RbxTree.html
//! [`RbxTree::insert_instance`]: struct.RbxTree.html#method.insert_instance
//! [`RbxTree::get_instance`]: struct.RbxTree.html#method.get_instance
//! [`RbxTree::get_instance_mut`]: struct.RbxTree.html#method.get_instance_mut
//! [`RbxInstanceProperties`]: struct.RbxInstanceProperties.html
//! [maplit]: https://crates.io/crates/maplit

mod unresolved_value;
mod brick_color;
mod id;
mod instance;
mod tree;
mod value;
mod value_conversion;

pub use crate::{
    unresolved_value::{AmbiguousRbxValue, UnresolvedRbxValue},
    brick_color::BrickColor,
    id::RbxId,
    instance::{RbxInstanceProperties, RbxInstance},
    tree::{RbxTree, Descendants},
    value::*,
};