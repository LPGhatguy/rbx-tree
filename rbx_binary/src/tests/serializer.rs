use rbx_dom_weak::{
    types::{Color3, Color3uint8, Ref, Region3, Vector3},
    InstanceBuilder, WeakDom,
};

use crate::{encode, text_deserializer::DecodedModel};

/// A basic test to make sure we can serialize the simplest instance: a Folder.
#[test]
fn just_folder() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder"));
    let mut buffer = Vec::new();

    encode(&tree, &[tree.root_ref()], &mut buffer).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that a tree containing some instances with a value and others
/// without will correctly fall back to (some) default value.
#[test]
fn partially_present() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder").with_children(vec![
        // This instance's `Value` property should be preserved.
        InstanceBuilder::new("StringValue").with_property("Value", "Hello"),
        // This instance's `Value` property should be the empty string.
        InstanceBuilder::new("StringValue"),
    ]));

    let root_refs = tree.root().children();

    let mut buffer = Vec::new();
    encode(&tree, root_refs, &mut buffer).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that unknown properties get serialized on instances.
#[test]
fn unknown_property() {
    let tree =
        WeakDom::new(InstanceBuilder::new("Folder").with_property("WILL_NEVER_EXIST", "Hi, mom!"));

    let mut buffer = Vec::new();
    encode(&tree, &[tree.root_ref()], &mut buffer).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that serializing a tree with an unimplemented property type returns
/// an error instead of panicking.
///
/// This test will need to be updated once we implement the type used here.
#[test]
fn unimplemented_type_known_property() {
    let tree = WeakDom::new(InstanceBuilder::new("UIListLayout").with_property(
        "Padding",
        Region3::new(Vector3::new(0.0, 0.0, 50.0), Vector3::new(0.0, 0.0, 50.0)),
    ));

    let mut buffer = Vec::new();
    let result = encode(&tree, &[tree.root_ref()], &mut buffer);

    assert!(result.is_err());
}

/// Ensures that serializing a tree with an unimplemented property type AND an
/// unknown property descriptor returns an error instead of panicking.
///
/// Because rbx_binary has additional logic for falling back to values with no
/// known property descriptor, we should make sure that logic works.
///
/// This test will need to be updated once we implement the type used here.
#[test]
fn unimplemented_type_unknown_property() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder").with_property(
        "WILL_NEVER_EXIST",
        Region3::new(Vector3::new(0.0, 0.0, 50.0), Vector3::new(0.0, 0.0, 50.0)),
    ));

    let mut buffer = Vec::new();
    let result = encode(&tree, &[tree.root_ref()], &mut buffer);

    assert!(result.is_err());
}

/// Ensures that the serializer returns an error instead of panicking if we give
/// it an ID not present in the tree.
#[test]
fn unknown_id() {
    let tree = WeakDom::new(InstanceBuilder::new("Folder"));

    let mut buffer = Vec::new();
    let result = encode(&tree, &[Ref::new()], &mut buffer);

    assert!(result.is_err());
}

/// Ensures that only one name for each logical property is serialized to a
/// file. Here, we use BasePart.Size and BasePart.size, which alias and both
/// serialize to BasePart.size.
///
/// For fun, we also have a part with no size property at all. It should default
/// to (4.0, 1.2, 2.0), a relic of Roblox's distant past.
#[test]
fn logical_properties_basepart_size() {
    let tree = WeakDom::new(
        InstanceBuilder::new("Folder")
            .with_child(
                InstanceBuilder::new("Part").with_property("Size", Vector3::new(1.0, 2.0, 3.0)),
            )
            .with_child(
                InstanceBuilder::new("Part").with_property("size", Vector3::new(4.0, 5.0, 6.0)),
            )
            .with_child(InstanceBuilder::new("Part")),
    );

    let mut buffer = Vec::new();
    encode(&tree, tree.root().children(), &mut buffer).expect("failed to encode model");

    let decoded = DecodedModel::from_reader(buffer.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}

/// Ensures that all valid combinations of color property names and
/// value types are properly handled.
#[test]
fn part_color() {
    let tree = WeakDom::new(
        InstanceBuilder::new("Folder")
            .with_child(
                InstanceBuilder::new("Part")
                    .with_property("Color3uint8", Color3::new(-0.25, 0.5, 1.2)),
            )
            .with_child(
                InstanceBuilder::new("Part")
                    .with_property("Color3uint8", Color3uint8::new(25, 86, 254)),
            )
            .with_child(
                InstanceBuilder::new("Part").with_property("Color", Color3::new(0.0, 0.5, 1.0)),
            )
            .with_child(
                InstanceBuilder::new("Part").with_property("Color", Color3uint8::new(1, 30, 100)),
            ),
    );

    let mut buf = Vec::new();
    let _ = encode(&tree, tree.root().children(), &mut buf);

    let decoded = DecodedModel::from_reader(buf.as_slice());
    insta::assert_yaml_snapshot!(decoded);
}
