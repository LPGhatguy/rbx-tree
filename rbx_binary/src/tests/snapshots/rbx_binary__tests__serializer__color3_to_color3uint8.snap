---
source: rbx_binary/src/tests/serializer.rs
expression: decoded
---
num_types: 1
num_instances: 3
chunks:
  - Inst:
      type_id: 0
      type_name: Part
      object_format: 0
      referents:
        - 0
        - 1
        - 2
  - Prop:
      type_id: 0
      prop_name: Color3uint8
      prop_type: Color3uint8
      values:
        - - 0
          - 128
          - 255
        - - 28
          - 85
          - 255
        - - 0
          - 128
          - 255
  - Prop:
      type_id: 0
      prop_name: Name
      prop_type: String
      values:
        - Part
        - Part
        - Part
  - Prnt:
      version: 0
      links:
        - - 0
          - -1
        - - 1
          - -1
        - - 2
          - -1
  - End
